### ISSUE TITLE
QEP 27 : Auxiliary storage

### ISSUE THREAD

**Issue by NathanW2** [qgis/QGIS-Enhancement-Proposals#27]
# 
# QGIS Enhancement 27 (was 20): Auxiliary storage

**Date** October 2015
**Author** Hugo Mercier
**Contact** hugo dot mercier at oslandia dot com
**Last Edited** May 2016
**Status**  
**Version**
## Summary

This proposal is about an evolution of QGIS aiming at storing auxiliary data in a layer. Such auxiliary data are data used mostly for the needs of QGIS (symbology) and have no real interest in being stored with the native raw geospatial data.

The need arises from the restrictions existing in the manual placement of labels.
Manual placement of labels are possible in QGIS by setting some labeling properties (X and Y position, and rotation angle optionaly) as being "data-defined", meaning that values come from a column (or an expression).
But setting this up on an existing layer requires either to add new columns to the source layer, while it is not always possible or desirable. Creating another table and using joins to it is possible but the synchronisation between the two tables is not possible, leading to a solution not satisfactory.

We propose here a more integrated solution where the creation of auxiliary columns or layers are brought more transparently to the user.
## Requirements

The proposed solution have the following requirements :
- It should allow to transparently add columns of any type to a layer, without adding them to the source data.
  Beside usual types (string, real, integer), we may also think of geometry types. We only consider types that can be converted to atomic types (string for WKT or WKB)
- The user should be able to add these columns and to edit values even if the layer is in read-only mode. Auxiliary data may be updated by the user without updating source data, or for read only source data.
- These values should not be stored with the source data, and should be carried by the project, i.e.
  when the project file is moved or copied, the auxiliary data should still be there.
- Adding a new row to the layer also adds a new row for each auxiliary field (with a null or default value)
- New GUI elements should ease the creation of auxiliary fields
  - on each "data-defined" property 
  - possibly on groups of "data-defined": for manual label placement or label connectors 
- It should be possible to save auxiliary data as regular data into a new vector layer.
## Proposed solution

We propose to extend the functionnalities of « joins » between vector layers by bringing write support and synchronisation.
Shortcuts in the main applications will allow to easily create these kind of linked layer with write support to answer the need of auxiliary fields in a layer.
## Core change : Project file format

The current project file format (XML) is too limited to carry an efficient storage of such auxiliary data. It should somehow be evolved to include a binary indexable format. SQLite is chosen as a first candidate for this data format. But it might evolve if performance issues are found.

Given the current state of the core code where lots of classes assume the project file is an XML file (everything that is saved to the project file has a writeXML method), it seems hard to completely refactor the project file format.

We propose to use a Zip file as the root of the new project file format. A Zip file exposes a filesystem-like structure.
The original XML-based project file will then be embedded in the Zip file. And additional files could de added.
It only adds a step after actually saving the XML tree: including it in a Zip file. Symetrically during the loading, the first step would be to extract the XML file and pass it to the existing readXML methods.

**Packaged layers**: In order to be able to embed layers in a project file, file-based providers will be changed to accept files contained in a zip file (through a url syntax like zip://zipfile.zip/dir/file for instance)

**Backwards Compatibility**: Project files from previous versions will still be available for opening. When saving back these project files, they will be saved in the new file format by default.
An option would be available to save a project in the old file format for compatibility. In this case, auxiliary fields could not be packaged in the project file, and it will be proposed to the user to extract them in the project file directory.
## Core change : Joined fields

Joins will be extended with write support. A joined field will be editable in the primary layer.

Two new options available during the creation of the join will be added :
- **delete cascade** option : when enabled, when a feature from the primary layer is deleted, the corresponding feature of the linked layer will be deleted as well
- **upsert on edit** option : when enabled each editing of the joined field from the primary layer will update the field of the corresponding field in the linked layer or insert a new feature in the linked layer if needed

Joins will also be extended to allow layers without explicit unique identifier field to be joined. In this case, the internal identifier of the feature will be used as a unique id. This feature will only be available for primary layers in read only mode.

During the creation of a join with write support, the user will be warned if a layer is joined more than once, in order to prevent concurrent write access by different layers which could lead to undesired behaviours.

**Edit modes**: Auxiliary fields should be editable even if the source layer is not. And edition of a joined field should be possible if the linked layer is in edit mode. QgsVectorLayer and QgsVectorLayerEditBuffer will be modified to implement these needs.
## GUI Changes

Data-defined properties are the most probable place where auxiliary fields will be used.

In the menu that can be found for each data-defined property, a new entry will be added to allow the creation, in a click, of an auxiliary field for that property and link it as the source of the data-defined property.

If no auxiliary field exists yet, the user could choose to create an auxiliary layer for a given layer. This will create a new sqlite layer joined to the source layer, stored as a file attachment to the zip file project.

Similar GUI shortcuts could then be added in order to automate the creation of a group of data-defined properties.

The creation of such auxiliary fields will also be transparent for the user: the manual labeling tools will be available and if no fields already exist for data-defined properties, they will be automatically created through auxiliary fields.

**Comment by anitagraser** [comment]
I like the idea!

Maybe I missed it, but will there be a way to copy the layer with auxiliary fields from one project to another?

**Comment by wonder-sk** [comment]
Hi Hugo! From me big +1 for the project files as ZIP files containing project XML file and extra files. This has been long on my to-do list as packaging extra stuff with project file would make it possible to have "pack'n'go" feature where the project files may bundle everything including layer data - or at least some external data (e.g. SVG/TTF files). It would be great if the new project file functionality would be generic - i.e. support any files, not just auxiliary fields - and would be also available to plugins that would be also allowed to read/write their extra files bundled in project file. In my view the auxiliary fields would be just one kind of extra file that could be bundled with project files... I think we will need a concept of "extra file handlers" which various components would register with project file in order to read/write those extra files.

By the way, I don't think there's any need to refactor project files from XML to SQLite storage - the stored data are mostly hierarchical, so XML in my opinion is more suited for that - and putting the project XML file as-is into a ZIP file sounds good.

I like the idea of having some simple means of extending layers - it will certainly make the advanced workflows easier.

I am not entirely sure if it is a good idea to have another mechanism for auxiliary fields - they _are_ outer left joins and it seems to me that would just add a lot of duplicated code to support auxiliary fields in addition to joined fields. Later in the text you mention in that joins do not support direct editing in main layer and synchronization with main layer - why not add support for that for joins instead of creating an alternative, yet very similar mechanism?

Could we please leave out the expression caching and custom project properties out of the QEP in order to stay focused on the key ideas?

**Comment by NathanW2** [comment]
Having a zip format for project files has also been a long time want of mine.  Something we can bundle and give to people would be perfect, even better as gdal supports loading files from ZIP so win!.

I'm a little confused about what is different to a virtual field as they are already working as expected and stored in the project file.

**Comment by haubourg** [comment]
My funder's two cents here:

- for the record @ccrook developped memory layer saver plugin that delegates auxiliary data to Qdatastream for similar needs. Limitations are - like anita points - that is not a standard format for ogr or QGIS, so some features do not work, like embed layers from other projects. We should have that work with auxiliary data. We use a lot plugins like import layers form project or layer's menu to copy layers definitions form one project to another. I bet that having auxiliary fields will end to breakage here. We have to find a gentle way to handle it, or maybe have a way to port auxiliary data from API commands. QLR could for instance be concerned too. It would be nice if it could embed auxiliary data, just as QGS (it's a subset of qgs maplayer definition).

- Having a key is mandatory to have it work to keep sync with datasource and auxiliary data. We have many users using shp, tab, csv, or db with no primary key set. How to handle that? I'm afraid we miss the target if we don't find a way for those much used formats. What about being able to define a primary key from within layer's definition? 

- I remember Mapinfo huge drawback with label, when technical ID of layers was recomputed on datasource's side. It's internal label data mixed lines, ending in beautiful crossing callouts maps. We might encounter that too. What about a "reset auxiliary data" feature to clean that up?  Anyway, this is a side edge case, on non critical data.

**Comment by mhugo** [comment]
@anitagraser Hmmm that's a good point. @haubourg talks about a plugin that allows to load a layer from another project. We probably should add something similar to this ... 

@wonder-sk I like the idea of having a generic project file where different components could read/write

> By the way, I don't think there's any need to refactor project files from XML to SQLite storage - the stored data are mostly hierarchical

SQLite was only for auxiliary fields. If it stays XML, what about indexes and geometries ?
Writing this, I realize there are of course two formats: the one on disk used for the loading/saving and the one in memory. Data in memory are written to disk when the project is saved. On disk it could stay as XML and in memory, it could be a custom format (like for memory layers) or whatever. Ok then :)

> Later in the text you mention in that joins do not support direct editing in main layer and synchronization with main layer - why not add support for that for joins instead of creating an alternative, yet very similar mechanism?

Yes, I am still a bit hesitant on the implementation to choose. I also prefer simple concepts and these are joins with triggers basically. I'll have to think more about pros and cons of both approaches

@NathanW2 
> I'm a little confused about what is different to a virtual field

Virtual fields are read-only, values defined by expressions. Here we need to store a value for each row

@haubourg 
> QLR could for instance be concerned too. It would be nice if it could embed auxiliary data, just as QGS

I don't understand this.

> What about being able to define a primary key from within layer's definition?

This is an evolution of each of the provider. And for simple cases (one-column primary keys) it does not seem very hard to add.

> I remember Mapinfo huge drawback with label, when technical ID of layers was recomputed on datasource's side

Yep. These are external data, so, by definition if IDs of the main table changes, problems happen. There is no magic here. But adding a consistency check on loading could be useful (if auxiliary IDs do not exist anymore for instance).

**Comment by haubourg** [comment]
>>QLR could for instance be concerned too. It would be nice if it could embed auxiliary data, just as QGS

>I don't understand this.

Sorry, I was too short on my explanations. QLR is a definition file. It saves maplayer style and reference to datasource. Opening a QLR will find the datasource and open it, the style it. Current Memory layers can't be opened that way. Auxiliary data described here should be . 

Have fun in Las Palmas for those concerned. ;-)

**Comment by mhugo** [comment]
Ahh. Got it :)
Yes QGis layer definition file could be extend to support auxiliary data. That's a good point. And it would be a way to share auxiliary data between projects.

**Comment by mhugo** [comment]
New section "alternative approach" just to add some matter on the idea of extending "joins".

**Comment by mhugo** [comment]
@wonder-sk About this idea of extending joins to become left joins on steroid with automatic update trigger and cascading, after thinking about it again (see the alternate approach section), I am not really fan.
I find it more complex to implement than a FeatureIterator that returns some more columns. And I am not sure there would be other uses of these left joins with synchronization apart from this particular need.

**Comment by mhugo** [comment]
Updated text to reflect one important thing: auxiliary fields should be editable even if the source layer is not.

**Comment by wonder-sk** [comment]
Hi @mhugo sorry for the lag...

Maybe I still do not understand how do you mean to implement the auxiliary fields, because the "auxiliary fields" approach and alternative approach seem still look the same... For example, what would be the difference in QgsVectorLayerFeatureIterator with aux fields compared to joined fields?

For the alternative approach I thought SQLite database would be used too (not a memory layer).

One more question to aux fields description:

> A new row is added to the auxiliary table when a new row has been added to the origin vector layer and a data has been set (not null).

Does this mean that for a layer there will always be a row in aux table even if it does not contain any extra data?

**Comment by mhugo** [comment]
The difference is from the user's point of view: either it's a new type of field (aux field) that the user can create on a layer and no more control is added (except exporting it), or it's a field that comes from a join defined on the layer with options enabled to make sure it synchronizes well.

In the first case, from the user's point of view everything is stored "somewhere with the project". In the second case, the creation of auxiliary fields is still transparent for the end user, but it relies on a join to a (memory/Sqlite) layer. Then the user could also choose to make its own join on its own data source. That exposes lots of new possibilities that should be carefully thought through, tested, and so on. I am not sure it is mandatory for the initial need.

Now speaking in terms of code, the two approaches will probably be very close in QgsVectorLayerFeatureIterator for instance.

> Does this mean that for a layer there will always be a row in aux table even if it does not contain any extra data?

"A new row is added to the auxiliary table when a new row has been added to the origin vector layer **and a data has been set**": I should rephrase it to make it more explicit. But the intent is to add a row only when it's needed.

**Comment by nyalldawson** [comment]
copying some comments which were lost from the original PR:

> New GUI elements should ease the creation of auxiliary fields
>  on each "data-defined" property
>  possibly on groups of "data-defined": for manual label placement or label connectors

This could just be done as required too. Eg, if user goes to move a label using the move label tool (and hasn't setup any existing data defined x/y), then the x/y aux fields are just instantly created and linked to the data defined property without any user intervention.

**Comment by nyalldawson** [comment]
>The current project file format (XML) is too limited to carry an efficient storage of such auxiliary data. It should somehow be evolved to include a binary indexable format. SQLite is chosen as a first candidate for this data format. But it might evolve if performance issues are found.

Also, where will the sqlite file be stored while a project is being edited? I assume the user will be working with a temporary version of the sqlite database which is only stored inside the project zip on project save?

**Comment by nyalldawson** [comment]
One other thing I would like to see included in this QEP is a discussion of the timeline and what will be done in regards to backwards compatibility.

I think 3.0 would be the ideal time for these changes to land, as changing project format mid way through a major version isn't very friendly. I know we are usually pretty lax with backwards compatibility of projects, but in general they still open in older versions (even if missing some styling/etc). Will there be facility to save a project in the current .qgs project format, for users who need to share projects with users who are stuck on the 2.x series?

**Comment by mhugo** [comment]
@nyalldawson 
Added a phrase about automatic creation of x/y fields.

The SQLite file will be used for serializing/deserializing on disk. In memory, it could be a copy of a previously sqlite file in a temp dir, or a ":mem:" sqlite. (Or even a kind of memory layer serialized to disk on save)

**Comment by mhugo** [comment]
Added something about project file compatibility the "other way" (saving a 3.0 project to a 2.x format): it would be possible, but without auxiliary fields saved

**Comment by mhugo** [comment]
Added a section on data integrity check.

**Comment by mhugo** [comment]
Slight updates to the text. @NathanW2 final draft for me

**Comment by mhugo** [comment]
In the hope of having this feature implemented one day, I will need to estimate the effort needed and get back to funders soon. Are there remaining objections or could two gentle developers spare me a +1 :) ?

**Comment by timlinux** [comment]
From my side +1 as I think the functionality describes will be very useful. A couple more questions from me (though not blockers for the QEP to be approved):

* What happens if the same layer is added more than once to the project - is the same aux file used?
* What happens with embedded layers from another project - do they carry over the aux data from the source project?

Looking forward to seeing this functionality coming to QGIS!

**Comment by mhugo** [comment]
@timlinux Thanks.
Embedded layers from another project (or imported layer from QLR) may carry auxiliary fields with them (see "Auxiliary data export" section). They will be imported as well.
Auxiliary fields are associated with a layer id, so that if you import twice the same layer, the two layers will have distinct aux fields, they won't be shared, they will be copied.

**Comment by mhugo** [comment]
Hi, I've reconsidered the approach suggested by @wonder-sk to extend joined layers by adding an edit mode with some synchronisation mechanism, and I finally find it better ... ;)
So the text has been updated to reflect these changes.
This should not change what is doable by the end user compared to the previous approach.

**Comment by mhugo** [comment]
So, that's a +1 by @timlinux. Anyone else for the second one ? @wonder-sk ? @nyalldawson ?

**Comment by wonder-sk** [comment]
For me this is +1

It is not explicitly mentioned in the QEP, but I assume the work will introduce API to allow other components of QGIS (or plugins) to insert / list / remove extra data files to/from the project, so it will not be limited just to the auxiliary layers. Is that correct?

(I am looking forward to the possibility to pack extra files like SVG or TTF in projects, so it is easier to pass around the project files)

**Comment by haubourg** [comment]
@wonder-sk yep! the idea is to free the path for embedding also color ramps, svg, logos images fonts, datasources for memory layer saver etc... 
That said, Memory layer saving would deserve a QEP itself to make a smooth roadmap for memory layser saver plugin users, and avoid caveat by saing all temporary layers coming from processing..

**Comment by wonder-sk** [comment]
@haubourg great! Agreed that saving memory layers would be a good topic for a separate QEP...

**Comment by haubourg** [comment]
Just a note to keep you informed that this work as started and is expected before 3.0 feature freeze. @pblottiere and @mhugo are working at the mine on that :)
Stay tuned to the PR's, we'll probably will have to discuss ergonomic choices there

**Comment by 3nids** [comment]
@haubourg can this be closed?

**Comment by haubourg** [comment]
Yes please!

**Comment by haubourg** [comment]
@NathanW2 can you close the QEP please? Thanks!

**Comment by nyalldawson** [comment]
I didn't think we were closing approved QEPs? I thought tagged as accepted = done

**Comment by NathanW2** [comment]
Yep sorry was a bit quick on the button. It's Monday remember :)
